# **DSCI 504: Intro to SQL**

## **Assignment 07 – Advanced PostgreSQL Functions**

**Symphony Hopkins**

**October 8, 2022**

**Part 1: Write a query that finds the total number of sales for the year 2015. Your query cannot use any aggregates. All calculations should be done using the data/time functions in this week's learning. Your output should be in two columns-- year\_sales and num\_of\_sales.**

In [13]:
SELECT DISTINCT(EXTRACT(YEAR FROM o.ord_date)) AS year_sales
      ,COUNT(ord_id) OVER (PARTITION BY EXTRACT(YEAR FROM o.ord_date)) AS num_of_sales
FROM dsci_504.orders AS o
WHERE EXTRACT(YEAR FROM o.ord_date) = 2015;

(1 row(s) affected)

Total execution time: 00:00:00.004

year_sales num_of_sales 2015 135

**Find the monthly sales for the same year use the same output as above.**

In [108]:
SELECT *
FROM
(SELECT DISTINCT(TO_CHAR(o.ord_date, 'YYYY-MM')) as year_sales
,COUNT(ord_id) OVER (PARTITION BY TO_CHAR(o.ord_date, 'YYYY-MM')) AS num_of_sales
FROM dsci_504.orders AS o) AS q1
WHERE q1.year_sales BETWEEN '2015-01' AND '2015-12'



(12 row(s) affected)

Total execution time: 00:00:00.034

year_sales num_of_sales 2015-01 12 2015-02 11 2015-03 12 2015-04 23 2015-05 10 2015-06 8 2015-07 10 2015-08 9 2015-09 9 2015-10 14 2015-11 7 2015-12 10

**Write a query to find the sales for all years in the data**

In [29]:
SELECT *
FROM
(SELECT DISTINCT(EXTRACT(YEAR FROM o.ord_date)) AS year_sales
      ,COUNT(ord_id) OVER (PARTITION BY EXTRACT(YEAR FROM o.ord_date)) AS num_of_sales
FROM dsci_504.orders AS o) AS q1
ORDER BY q1.year_sales

(20 row(s) affected)

Total execution time: 00:00:00.008

year_sales num_of_sales 2000 150 2001 137 2002 123 2003 112 2004 119 2005 131 2006 154 2007 124 2008 122 2009 129 2010 138 2011 131 2012 126 2013 132 2014 113 2015 135 2016 133 2017 134 2018 135 2019 121

- **How many sales were there in 2010?**
    - 138

**Write a query to track customer acquisition in 2018. Display the number of customers acquired each month. Your output should be limited to two columns. You can name the columns as you see fit.**

In [32]:
SELECT *
FROM 
(SELECT DISTINCT(TO_CHAR(c.cus_join_date, 'YYYY-MM')) as year_month_cus
,COUNT(c.cus_id) OVER (PARTITION BY TO_CHAR(c.cus_join_date, 'YYYY-MM')) AS num_of_cus
FROM dsci_504.customers AS c) AS q1
WHERE q1.year_month_cus BETWEEN '2018-01' AND '2018-12'
ORDER BY q1.year_month_cus

(12 row(s) affected)

Total execution time: 00:00:00.004

year_month_cus num_of_cus 2018-01 7 2018-02 10 2018-03 11 2018-04 8 2018-05 13 2018-06 8 2018-07 6 2018-08 14 2018-09 6 2018-10 9 2018-11 17 2018-12 13

**Write the same query to track customer acquisition over the entire dataset. You can customize your output to your liking.**

In [34]:
SELECT *
FROM
(SELECT DISTINCT(TO_CHAR(c.cus_join_date, 'YYYY')) as year_cus
,COUNT(c.cus_id) OVER (PARTITION BY TO_CHAR(c.cus_join_date, 'YYYY')) AS num_of_cus
FROM dsci_504.customers AS c) AS q1
ORDER BY q1.year_month_cus

(20 row(s) affected)

Total execution time: 00:00:00.007

year_month_cus num_of_cus 2001 137 2002 122 2003 153 2004 124 2005 151 2006 146 2007 144 2008 127 2009 149 2010 131 2011 145 2012 135 2013 132 2014 119 2015 132 2016 134 2017 155 2018 122 2019 140 2020 21

- **Which year had the most customers?** 
    - 2017

**Write a query to find the day of week the customer with id 1800 joined. Output the customer's table ID, customer ID, lastname and day of the week they joined.**

In [41]:
SELECT c.cus_id
      ,c.cus_last_name 
      ,to_char(c.cus_join_date, 'Day') AS joined_weekday
FROM dsci_504.customers AS c
WHERE c.cus_id = 1800;


(1 row(s) affected)

Total execution time: 00:00:00.002

cus_id cus_last_name joined_weekday 1800 Dickerson Thursday

- **What is the last name of customer 1800 and row of the week customer 1800 joined?**
    - Last Name = Dickerson, Row Num = 492

In [48]:
SELECT * 
FROM
(SELECT c.cus_id
      ,c.cus_last_name 
      ,to_char(c.cus_join_date, 'Day') AS joined_weekday
      ,ROW_NUMBER() OVER () as row_num
FROM dsci_504.customers AS c) AS q1
WHERE q1.cus_id = 1800;


(1 row(s) affected)

Total execution time: 00:00:00.005

cus_id cus_last_name joined_weekday row_num 1800 Dickerson Thursday 492

**Part 2: Determine the number of days between a customer joining and the delivery of their order. Limit your query to the year 2018.**

In [80]:
/* These results are very strange results because if a customer joined in 2018, I don't know how they have an order ship date
from 2007. So the customer join dates, the ship dates, or query isn't correct.*/
SELECT *
FROM
(SELECT c.cus_id
      ,c.cus_last_name
      ,c.cus_join_date
      ,o.ord_ship_date
      ,c.cus_join_date::DATE - o.ord_ship_date::DATE AS num_of_days
FROM dsci_504.customers AS c
INNER JOIN dsci_504.orders AS o
ON c.cus_id = o.cus_id
WHERE c.cus_join_date BETWEEN '2018-01-01' AND '2018-12-31') AS q1;



(122 row(s) affected)

Total execution time: 00:00:00.007

cus_id cus_last_name cus_join_date ord_ship_date num_of_days 1279 Connery 2018-09-23 2007-02-24 4229 581 Dickerson 2018-12-11 2006-05-07 4601 407 Hart 2018-08-08 2006-11-02 4297 164 Jones 2018-05-02 2007-12-14 3792 1371 Shelton 2018-04-23 2016-09-18 582 378 Seiler 2018-04-25 2014-03-24 1493 1446 Ramirez 2018-06-10 2003-07-09 5450 1475 Delacruz 2018-06-02 2001-10-21 6068 1477 Gorman 2018-03-15 2019-04-24 -405 1495 Ridgley 2018-11-01 2016-05-11 904 1573 Rice 2018-03-21 2004-06-06 5036 1578 Goode 2018-08-21 2009-03-25 3436 919 Gonzales 2018-03-14 2018-04-28 -45 746 Ridgley 2018-01-04 2006-08-13 4162 1741 Slain 2018-03-25 2011-04-11 2540 1744 Jones 2018-11-17 2003-12-02 5464 1761 Barrister 2018-01-25 2009-05-23 3169 1783 Stevens 2018-10-19 2016-12-07 681 1800 Dickerson 2018-10-11 2008-07-21 3734 1862 Maurer 2018-03-16 2001-06-22 6111 1872 Johnasen 2018-05-25 2017-06-26 333 1873 Delacruz 2018-02-07 2016-06-25 592 1918 Smith-Warren 2018-05-28 2017-07-18 314 1941 Acosta 2018-07-13 2018-06-29 14 1944 Brooks 2018-08-25 2015-04-17 1226 2041 Delacruz 2018-10-06 2003-09-28 5487 2082 Dickerson 2018-04-28 2013-01-10 1934 2087 Aaronson 2018-05-23 2011-01-29 2671 761 Calahan 2018-04-16 2012-11-25 1968 2147 Booker 2018-03-11 2012-05-29 2112 2148 Sanchez 2018-05-01 2002-11-27 5634 2155 Thomason 2018-03-01 2011-08-29 2376 2216 Brooks 2018-07-13 2019-10-08 -452 563 Dickerson 2018-12-07 2000-08-05 6698 730 Manchester 2018-05-22 2003-12-07 5280 2237 Johnasen 2018-09-25 2019-05-15 -232 2258 Thomason 2018-01-29 2006-11-07 4101 2266 Gonzales 2018-01-07 2010-03-28 2842 2268 Sherman 2018-11-01 2014-06-11 1604 770 Presten 2018-03-01 2013-12-06 1546 2298 Thomason 2018-03-22 2012-10-16 1983 2368 Presten 2018-02-09 2003-07-01 5337 2371 Slain 2018-10-19 2015-03-01 1328 2376 Seiler 2018-05-19 2012-08-11 2107 1105 Maurer 2018-06-14 2018-01-25 140 2441 Aaronson 2018-06-06 2007-12-03 3838 2482 Rowley 2018-10-23 2001-06-22 6332 2485 Jones 2018-06-25 2006-12-25 4200 183 Smith 2018-08-10 2018-07-28 13 2529 Banks 2018-11-21 2006-04-28 4590 2578 Jackson 2018-12-05 2018-08-08 119 109 Jones 2018-08-25 2008-11-24 3561 539 French 2018-10-20 2004-06-11 5244 124 Deleon 2018-08-16 2019-05-14 -271 110 Moore 2018-08-23 2009-05-03 3399 114 French 2018-07-30 2007-11-19 3906 1151 Jackson 2018-12-19 2015-11-06 1139 155 Cooper 2018-05-09 2013-04-10 1855 1879 Gonzales 2018-11-11 2006-02-05 4662 2598 Rowley 2018-11-01 2017-02-12 627 78 Acosta 2018-06-05 2015-04-14 1148 2487 French 2018-10-20 2001-11-10 6188 2145 Ramirez 2018-06-24 2002-02-15 5973 267 Jones 2018-12-24 2003-01-13 5824 1167 Etzler 2018-11-27 2019-07-10 -225 1223 Conley 2018-11-29 2003-03-25 5728 786 Smith 2018-08-20 2001-05-18 6303 714 Etzler 2018-04-08 2005-04-21 4735 832 Banks 2018-08-15 2008-11-20 3555 431 Smith-Warren 2018-02-07 2011-12-26 2235 445 Maurer 2018-07-06 2013-08-14 1787 447 Peters 2018-11-21 2007-08-26 4105 472 Jackson 2018-07-12 2015-10-06 1010 47 Manchester 2018-12-14 2012-11-02 2233 350 Deleon 2018-10-01 2017-10-15 351 977 Jordan 2018-02-23 2003-08-11 5310 193 Ellis 2018-01-16 2003-07-25 5289 46 Dickerson 2018-02-19 2001-03-01 6199 1292 Dickerson 2018-09-03 2008-06-05 3742 185 Reynolds 2018-11-18 2007-11-20 4016 204 Calahan 2018-12-30 2009-03-21 3571 1181 Barrister 2018-05-08 2008-04-25 3665 978 Jordan 2018-02-15 2012-09-13 1981 1193 Dewey 2018-11-03 2010-12-26 2869 295 MacBain 2018-08-21 2000-03-15 6733 595 Reynolds 2018-02-25 2012-09-23 1981 492 Etzler 2018-03-19 2015-01-09 1165 527 Robinson 2018-02-23 2016-07-11 592 537 Maurer 2018-05-23 2014-09-22 1339 982 Gorman 2018-11-04 2010-01-08 3222 627 Presley 2018-11-26 2003-12-04 5471 640 Conley 2018-03-15 2013-11-06 1590 641 Rice 2018-07-05 2000-05-12 6628 815 Saylor 2018-08-02 2009-08-04 3285 2287 Garcia 2018-01-18 2014-10-31 1175 853 Brooks 2018-09-19 2000-07-01 6654 1246 Sherman 2018-06-18 2010-10-13 2805 1614 Banks 2018-11-11 2010-08-12 3013 1650 Barrister 2018-04-03 2004-03-31 5116 1459 Acosta 2018-02-17 2012-06-03 2085 2596 Gorman 2018-08-10 2017

- **What was the shortest and longest delay between acquisition and delivery in 2016?**
    - As stated before, these are very strange numbers because if a person joined in 2016, they shouldn't have any order shipped from 2007. 
        - The shortest delay was -1445 days (?)
        - The longest delay was 5941 (?).

In [84]:
SELECT *
FROM
(SELECT c.cus_id
      ,c.cus_last_name
      ,c.cus_join_date
      ,o.ord_ship_date
      ,c.cus_join_date::DATE - o.ord_ship_date::DATE AS num_of_days
      ,min(c.cus_join_date::DATE - o.ord_ship_date::DATE) over () as min_days
      ,max(c.cus_join_date::DATE - o.ord_ship_date::DATE) over () as max_days
FROM dsci_504.customers AS c
INNER JOIN dsci_504.orders AS o
ON c.cus_id = o.cus_id
WHERE c.cus_join_date BETWEEN '2016-01-01' AND '2016-12-31') AS q1;


(134 row(s) affected)

Total execution time: 00:00:00.008

cus_id cus_last_name cus_join_date ord_ship_date num_of_days min_days max_days 1242 Smith-Warren 2016-07-20 2007-06-20 3318 -1445 5941 1255 Cooper 2016-03-27 2018-09-11 -898 -1445 5941 1282 Acosta 2016-02-02 2002-07-01 4964 -1445 5941 1301 Presten 2016-01-25 2017-10-28 -642 -1445 5941 1228 Stevens 2016-05-26 2011-02-15 1927 -1445 5941 1338 French 2016-05-10 2011-07-25 1751 -1445 5941 1781 Smith 2016-03-08 2014-11-17 477 -1445 5941 735 Manning 2016-11-15 2015-04-08 587 -1445 5941 1390 Applewood 2016-09-22 2000-06-19 5939 -1445 5941 149 Deleon 2016-10-06 2006-04-08 3834 -1445 5941 1406 Connery 2016-11-12 2011-12-07 1802 -1445 5941 1438 Moore 2016-12-30 2017-03-03 -63 -1445 5941 1469 Ramirez 2016-01-22 2005-11-12 3723 -1445 5941 1483 Jamieson 2016-09-22 2004-07-31 4436 -1445 5941 1489 Delacruz 2016-04-22 2010-02-06 2267 -1445 5941 739 Jones 2016-03-25 2013-07-09 990 -1445 5941 549 Applewood 2016-09-19 2017-10-25 -401 -1445 5941 1623 Ridgley 2016-03-09 2005-07-20 3885 -1445 5941 1624 Smith 2016-07-21 2014-12-28 571 -1445 5941 1630 Conley 2016-07-15 2017-10-13 -455 -1445 5941 745 Hart 2016-09-25 2017-08-31 -340 -1445 5941 1752 Delapena 2016-03-23 2009-06-17 2471 -1445 5941 750 Jamieson 2016-01-15 2005-11-25 3703 -1445 5941 1769 Manning 2016-05-07 2000-01-31 5941 -1445 5941 346 Tankersly 2016-05-28 2015-08-26 276 -1445 5941 1817 Ramirez 2016-06-19 2011-02-13 1953 -1445 5941 1842 Roberts 2016-11-29 2013-01-09 1420 -1445 5941 1867 Robinson 2016-05-10 2002-09-21 4980 -1445 5941 1876 Aaronson 2016-01-07 2019-12-22 -1445 -1445 5941 1929 Deleon 2016-01-09 2001-01-10 5477 -1445 5941 160 Dewey 2016-11-27 2017-06-28 -213 -1445 5941 2036 Sherman 2016-09-25 2017-06-23 -271 -1445 5941 2059 Smith 2016-01-13 2009-12-31 2204 -1445 5941 2077 Manchester 2016-07-07 2017-09-22 -442 -1445 5941 558 Jamieson 2016-12-29 2003-10-12 4827 -1445 5941 2118 Jones 2016-11-23 2009-12-01 2549 -1445 5941 2134 Slain 2016-06-20 2010-09-30 2090 -1445 5941 2135 Presley 2016-05-01 2006-09-12 3519 -1445 5941 2154 Bangeson 2016-02-29 2019-01-05 -1041 -1445 5941 2165 Robinson 2016-05-21 2007-11-11 3114 -1445 5941 2201 Harper 2016-01-28 2017-08-27 -577 -1445 5941 2228 Sanchez 2016-06-16 2004-01-31 4520 -1445 5941 2234 Connery 2016-09-03 2000-06-06 5933 -1445 5941 2242 Smith-Warren 2016-05-03 2004-03-14 4433 -1445 5941 2251 Dawson 2016-09-02 2016-02-11 204 -1445 5941 2262 Aaronson 2016-09-16 2011-06-22 1913 -1445 5941 2274 Riochardson 2016-08-01 2005-01-12 4219 -1445 5941 2297 Applewood 2016-03-08 2014-08-05 581 -1445 5941 2359 Rice 2016-01-19 2003-08-21 4534 -1445 5941 2365 Tompsen 2016-08-31 2019-10-24 -1149 -1445 5941 2374 Slain 2016-09-04 2002-12-30 4997 -1445 5941 2385 Delacruz 2016-04-08 2013-09-06 945 -1445 5941 2392 Riochardson 2016-09-05 2015-06-13 450 -1445 5941 2399 Delapena 2016-02-01 2011-09-21 1594 -1445 5941 2414 Rowley 2016-02-26 2004-06-19 4269 -1445 5941 2416 Delacruz 2016-01-03 2002-10-24 4819 -1445 5941 2427 Calahan 2016-07-07 2006-01-03 3838 -1445 5941 2429 Slain 2016-06-26 2012-06-30 1457 -1445 5941 8 Jones 2016-09-24 2003-04-01 4925 -1445 5941 2478 Manchester 2016-12-10 2008-09-03 3020 -1445 5941 2495 Tankersly 2016-04-08 2009-06-11 2493 -1445 5941 2514 Ellis 2016-07-25 2010-12-19 2045 -1445 5941 2515 Monk 2016-12-27 2004-05-23 4601 -1445 5941 2519 Dewey 2016-07-23 2003-01-30 4923 -1445 5941 2528 Rowley 2016-05-26 2013-08-09 1021 -1445 5941 104 Delacruz 2016-04-27 2005-02-10 4094 -1445 5941 1226 Harper 2016-05-14 2009-11-01 2386 -1445 5941 2543 Ridgley 2016-08-30 2005-01-23 4237 -1445 5941 970 Harper 2016-12-22 2012-03-16 1742 -1445 5941 2588 Dawson 2016-02-25 2017-04-22 -422 -1445 5941 2591 Dawson 2016-11-30 2008-01-09 3248 -1445 5941 1946 Presten 2016-09-09 2018-10-12 -763 -1445 5941 405 Reynolds 2016-03-12 2006-06-13 3560 -1445 5941 901 Booker 2016-09-28 2016-12-21 -84 -1445 5941 24 Smith 2016-02-10 2003-09-27 4519 -1445 5941 244 Peters 2016-09-13 2017-04-21 -220 -1445 5941 39 Sherman 2016-11-22 2002-03-09 5372 -1445 5941 399 Riochardson 2016-01-09 

**Part 3: Perform a query to find all customers with orders fulfilled through the Dallas warehouse. This should be a simple query with no joins. Order your results by the total value of each customer's order.**

In [91]:
--Dallas warehouse code is 2.
SELECT *
FROM  
(SELECT o.cus_id
      ,o.warehouse_id
      ,o.order_tot
      ,SUM(o.order_tot) OVER(PARTITION BY o.warehouse_id ORDER BY o.order_tot) AS cus_order_tot
FROM dsci_504.orders AS o) AS q1
WHERE q1.warehouse_id = 2;

(859 row(s) affected)

Total execution time: 00:00:00.023

cus_id warehouse_id order_tot cus_order_tot 959 2 1850 5550 2318 2 1850 5550 1572 2 1850 5550 770 2 1924 9398 2188 2 1924 9398 160 2 1932.33 11330.33 2242 2 1937.88 13268.21 268 2 1949.99 21068.17 2306 2 1949.99 21068.17 2549 2 1949.99 21068.17 2048 2 1949.99 21068.17 1583 2 1951.75 24971.67 1737 2 1951.75 24971.67 285 2 1961 30854.67 458 2 1961 30854.67 2468 2 1961 30854.67 2419 2 1970.25 32824.92 1328 2 1977.19 34802.11 349 2 1984.13 36786.24 2457 2 2066.99 38853.23 1150 2 2073.81 47148.47 811 2 2073.81 47148.47 1247 2 2073.81 47148.47 309 2 2073.81 47148.47 540 2 2084.05 51316.57 297 2 2084.05 51316.57 1367 2 2199 57913.57 1903 2 2199 57913.57 167 2 2199 57913.57 1437 2 2296.86 60210.43 755 2 2305.65 62516.08 262 2 2319.95 69475.93 1776 2 2319.95 69475.93 750 2 2319.95 69475.93 2116 2 2325.44 71801.37 1242 2 2330.94 78794.19 152 2 2330.94 78794.19 1827 2 2330.94 78794.19 27 2 2338.64 81132.83 636 2 2341.94 85816.71 1208 2 2341.94 85816.71 2033 2 2352.93 88169.64 287 2 2639 98725.64 248 2 2639 98725.64 2028 2 2639 98725.64 267 2 2639 98725.64 300 2 2744.56 101470.20 739 2 2784.15 107038.50 968 2 2784.15 107038.50 2156 2 2797.34 109835.84 2535 2 2799 115433.84 2254 2 2799 115433.84 1112 2 2820.43 118254.27 2144 2 2823.73 123901.73 2550 2 2823.73 123901.73 2170 2 2830.33 126732.06 157 2 2923.56 129655.62 1281 2 2959.94 132615.56 2597 2 2966.94 141516.38 1323 2 2966.94 141516.38 1089 2 2966.94 141516.38 358 2 2976.74 144493.12 1982 2 2980.94 147474.06 2184 2 2994.93 150468.99 1720 2 3001.93 153470.92 905 2 3064 159598.92 527 2 3064 159598.92 2271 2 3129.36 165857.64 502 2 3129.36 165857.64 977 2 3138.39 168996.03 832 2 3154.94 175305.91 1535 2 3154.94 175305.91 1026 2 3174.5 184829.41 1590 2 3174.5 184829.41 1226 2 3174.5 184829.41 1023 2 3186.56 194389.09 5 2 3186.56 194389.09 1825 2 3186.56 194389.09 483 2 3189.54 203957.71 1724 2 3189.54 203957.71 1439 2 3189.54 203957.71 2539 2 3199 213554.71 90 2 3199 213554.71 783 2 3199 213554.71 2393 2 3200.35 216755.06 1947 2 3204.59 219959.65 744 2 3209.08 223168.73 2566 2 3212.6 226381.33 1609 2 3219.63 232820.59 4 2 3219.63 232820.59 2354 2 3232.52 236053.11 1133 2 3247.84 242548.79 1012 2 3247.84 242548.79 1699 2 3258.56 245807.35 2586 2 3274.65 249082.00 836 2 3326.96 259062.88 1211 2 3326.96 259062.88 2384 2 3326.96 259062.88 1151 2 3337.44 269075.20 1051 2 3337.44 269075.20 1707 2 3337.44 269075.20 2197 2 3354.15 272429.35 445 2 3378.14 275807.49 1233 2 3382.94 279190.43 2594 2 3390.94 292754.19 1852 2 3390.94 292754.19 2366 2 3390.94 292754.19 2011 2 3390.94 292754.19 2065 2 3393.6 296147.79 1448 2 3399 299546.79 1698 2 3401.62 309751.65 1129 2 3401.62 309751.65 15 2 3401.62 309751.65 2381 2 3406.94 313158.59 1764 2 3417.67 323411.60 479 2 3417.67 323411.60 1514 2 3417.67 323411.60 2403 2 3429.7 326841.30 721 2 3433.72 333708.74 1563 2 3433.72 333708.74 1710 2 3441.74 337150.48 1712 2 3500 351150.48 1161 2 3500 351150.48 858 2 3500 351150.48 700 2 3500 351150.48 1057 2 3504.8 354655.28 1755 2 3530.08 358185.36 1413 2 3533.45 365252.26 2595 2 3533.45 365252.26 2259 2 3534.96 372322.18 692 2 3534.96 372322.18 667 2 3558.72 375880.90 2423 2 3560.45 379441.35 1149 2 3563.85 390132.90 1853 2 3563.85 390132.90 2415 2 3563.85 390132.90 1174 2 3572.2 400849.50 2087 2 3572.2 400849.50 2382 2 3572.2 400849.50 2590 2 3585.95 408021.40 993 2 3585.95 408021.40 1048 2 3594.44 411615.84 1386 2 3600 418815.84 1521 2 3600 418815.84 433 2 3601.69 422417.53 77 2 3602.94 426020.47 1848 2 3605.9 436838.17 1018 2 3605.9 436838.17 2532 2 3605.9 436838.17 2305 2 3614.33 447681.16 2097 2 3614.33 447681.16 143 2 3614.33 447681.16 1677 2 3620.23 451301.39 798 2 3632.68 454934.07 1351 2 3645.43 462224.93 2265 2 3645.43 462224.93 542 2 3666.25 465891.18 2079 2 3669.75 469560.93 607 2 3672.44 476905.81 111 2 3672.44 476905.81 928 2 3701.25 480607.06 385 2 3707.25 484314.31 1862 2 3710 491734.31 205 2 3710 491734.31 643 2 3720.31 495454.62 2330 2 3722.25 502899.12 920 2 3722.25 502899.12 2480 2 3724.66 50662

**Part 4: Shimano has reached out to OPC and has asked that OPC provide a listing of all customers that purchased a bike with a combination of Shimano's race components XTR due to a critical recall issuance. They have concerns that not all purchasers have registered their bikes or that they may be out of warranty. Write a query that outputs the following elements to a table that can be exported and provided to Shimano. The table should be stored in the public schema.**

- **Customer ID**
- **Customer Last Name**
- **Customer First Name**
- **Customer State**
- **Customer Phone Number**

In [102]:
--Finding builds containing Shimano components

SELECT *
FROM dsci_504.components AS c
WHERE c.comp_name LIKE '%Shimano XTR%';

--Shimano component IDs: 66, 214, 225, 229

SELECT *
FROM dsci_504.builds AS b
WHERE b.comp_brake = '66'
OR b.comp_shifter = '214'
OR b.comp_cassette = '225'
OR b.comp_derail = '229'
OR b.comp_chain_ring ='274';

--Bike build IDs with Shimano components: 45, 10, 22

--Creating customers with Shimano XTR bike components.

CREATE TABLE public.shimano_recall AS 
SELECT c.cus_id 
      ,c.cus_last_name
      ,c.cus_first_name
      ,c.cus_state
      ,c.cus_phone
FROM dsci_504.customers AS c
INNER JOIN dsci_504.orders AS o
ON c.cus_id = o.cus_id 
WHERE o.build_id IN (45, 10, 22);




(5 row(s) affected)

(3 row(s) affected)

Commands completed successfully

Total execution time: 00:00:00.028

comp_id comp_name comp_cost comp_supplier comp_cat sup_id 66 Shimano XTR Trail 275 44 brake NULL 214 Shimano XTR 137 44 shifter NULL 225 Shimano XTR M9100 389.99 44 cassette NULL 229 Shimano XTR M9100 180 44 derailleur NULL 274 Shimano XTR 78 44 chain_ring NULL

build_id comp_hbar comp_brake comp_wheels comp_fork comp_shock comp_shifter comp_cassette comp_derail comp_dropper comp_crank comp_chain_ring build_name 45 26 67 72 123 192 273 274 229 259 273 273 megalodon 10 12 61 78 136 202 206 219 229 246 264 271 wild ride 22 17 66 106 134 184 274 274 229 241 274 274 trail four

- **What is the customer ID of the customer on line 20?**
    
    - 1900

In [106]:
SELECT *
FROM 
(SELECT *
       ,ROW_NUMBER() OVER() AS row_num
 FROM public.shimano_recall) as q1
WHERE q1.row_num = 20

(1 row(s) affected)

Total execution time: 00:00:00.003

cus_id cus_last_name cus_first_name cus_state cus_phone row_num 1900 Harper Cooper 11 9269844955 20

**Part 5: Write a query that will aggregate the total of all bikes sold. Group the output by the bike manufacturer. Your output should contain the following elements:**

- **Bike Manufacturer**
- **Bike Name**
- **Bike Description**
- **Bike Country of Origin**
- **Total of all sales**

**Sort your output by the sum of the total order, manufacturer, and the bike name**

In [2]:
SELECT * 
FROM 
(SELECT p.prod_manufacturer
       ,p.prod_name
       ,p.prod_description
       ,p.country_origin 
       ,COUNT(o.prod_id) OVER (PARTITION BY p.prod_manufacturer) AS tot_sales
FROM dsci_504.products AS p
INNER JOIN dsci_504.orders AS o 
ON p.prod_id = o.prod_id) AS q1 
ORDER BY q1.tot_sales, q1.prod_manufacturer, q1.prod_name;

(2599 row(s) affected)

Total execution time: 00:00:00.069

prod_manufacturer prod_name prod_description country_origin tot_sales 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 70 Troy Full-Suspension Mountain Bike 2 40 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 66 Firebird Full-Suspension Mountain Bike 2,3,4 42 65 Rail Full-Suspension Mountain Bike 2,3 45 65 Rail Full-Suspension

- **What was the bike with the most sales?**
    
    - <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">The bike with the most sales was SB160.</span>
- **What was the total of sales for the Nukeproof Scout?**
    
    - The total sales for the Nukeproof Scout was 126.

**Write a query to find the total value of all bikes returned by customers. Your output should include the basic customer info, amount of return, and date of return. You can coordinate the output as you see fit but be sure to not include any PII data.**

In [110]:
--Finding total return value by customer.
SELECT c.cus_id
      ,c.cus_last_name 
      ,o.ord_date
      ,r.rac_id
      ,SUM(r.tot_ret_amnt) OVER (PARTITION BY r.cus_id ORDER BY r.tot_ret_amnt) AS tot_ret_amnt
FROM dsci_504.customers AS c
INNER JOIN dsci_504.racs AS r 
ON c.cus_id = r.cus_id 
INNER JOIN dsci_504.orders AS o 
ON o.ord_id = r.ord_id;

--Finding highest amount returned for second question.
SELECT MAX(q1.tot_ret_amnt) AS max_ret_amnt
FROM
(SELECT c.cus_id
      ,c.cus_last_name 
      ,o.ord_date
      ,r.rac_id
      ,SUM(r.tot_ret_amnt) OVER (PARTITION BY  r.cus_id ORDER BY r.tot_ret_amnt) AS tot_ret_amnt
FROM dsci_504.customers AS c
INNER JOIN dsci_504.racs AS r 
ON c.cus_id = r.cus_id 
INNER JOIN dsci_504.orders AS o 
ON o.ord_id = r.ord_id) AS q1;

(1024 row(s) affected)

(1 row(s) affected)

Total execution time: 00:00:00.059

cus_id cus_last_name ord_date rac_id tot_ret_amnt 1 Seiler 2010-10-12 950 0 3 Roberts 2013-04-28 375 368 4 Smith-Warren 2005-09-29 77 0 4 Smith-Warren 2001-09-30 816 341 8 Jones 2000-08-26 504 0 15 Sherman 2007-07-23 880 230 16 Manning 2017-07-21 156 289 18 Brooks 2017-12-08 258 657 19 Barrister 2000-09-17 625 233 20 Peters 2004-08-27 229 581 21 Jones 2000-03-23 307 0 22 Ridgley 2005-10-17 1001 0 24 Smith 2019-06-08 396 0 24 Smith 2007-02-18 479 16 24 Smith 2003-09-19 24 429 25 Riochardson 2005-05-06 687 0 29 Saylor 2011-09-09 964 0 29 Saylor 2002-05-01 95 247 29 Saylor 2019-03-09 9 1203 30 Tompsen 2001-10-05 473 0 34 Peters 2011-01-30 345 0 34 Peters 2000-06-30 853 0 37 Delacruz 2006-07-05 423 294 37 Delacruz 2012-11-26 63 765 38 Jackson 2002-12-17 700 622 42 MacBain 2015-04-28 629 0 42 MacBain 2017-08-26 745 769 46 Dickerson 2002-10-01 702 0 47 Manchester 2011-12-10 158 0 48 Booker 2016-03-30 741 0 54 Deleon 2011-05-09 607 615 56 Garth 2016-06-28 894 0 57 Rowley 2016-07-18 577 0 57 Rowley 2006-03-16 829 223 62 Deleon 2017-08-02 430 0 64 Ellis 2001-08-01 846 0 65 Smith-Warren 2019-08-12 248 581 65 Smith-Warren 2001-11-02 145 1190 70 Sanchez 2017-11-06 665 100 72 Rice 2015-05-17 269 320 73 Garcia 2009-10-02 138 211 78 Acosta 2016-07-27 395 0 78 Acosta 2006-10-07 262 0 82 Archibald 2001-08-05 809 0 84 Monk 2018-08-08 27 0 85 Stevens 2006-05-14 834 1044 86 Etzler 2015-04-07 208 0 91 Delacruz 2017-06-04 877 1041 92 Banks 2011-09-13 676 672 93 Johnasen 2006-04-03 149 0 95 Manchester 2012-02-14 163 0 95 Manchester 2005-10-16 762 387 98 French 2015-03-23 904 0 99 Manning 2004-07-15 955 1059 104 Delacruz 2004-01-04 139 0 105 Jones 2002-03-12 53 965 109 Jones 2015-12-25 817 1041 110 Moore 2003-06-21 279 405 111 Reynolds 2008-03-09 482 247 111 Reynolds 2008-06-14 938 600 112 Thomason 2018-08-28 308 260 116 Jones 2007-03-25 463 0 120 French 2016-12-13 901 0 125 Robinson 2018-07-06 659 353 126 Williamson 2002-10-13 435 0 128 Cooper 2003-06-20 613 0 130 Moore 2002-11-24 466 0 131 Rice 2000-10-05 693 0 133 Thomason 2004-05-30 864 0 137 Connery 2002-09-05 962 0 140 Monk 2001-02-24 46 0 140 Monk 2005-04-12 714 0 146 Shelton 2002-03-09 424 0 150 Conley 2008-11-08 11 0 153 Riochardson 2004-07-22 271 0 153 Riochardson 2000-03-25 408 857 155 Cooper 2009-09-06 41 615 157 Manchester 2000-01-09 885 236 157 Manchester 2006-04-10 525 1330 162 Willis 2014-06-22 252 0 162 Willis 2010-03-09 132 149 163 Smith 2008-11-05 422 197 164 Jones 2002-11-28 443 976 170 Jackson 2007-04-22 836 0 172 Etzler 2012-10-17 862 421 176 Tompsen 2016-08-03 998 0 178 Williamson 2005-11-14 591 0 181 Johnasen 2012-11-10 32 0 184 Ridgley 2000-11-19 729 937 187 Jordan 2011-06-30 253 767 188 Dickerson 2009-12-24 726 98 188 Dickerson 2005-05-11 910 582 189 Saylor 2009-10-07 928 941 191 Delacruz 2009-05-13 598 0 191 Delacruz 2007-07-25 889 1103 196 Seiler 2007-10-08 43 955 199 Jamieson 2015-10-26 616 0 199 Jamieson 2007-03-31 811 335 207 French 2016-04-22 1000 227 210 Ellis 2014-06-27 593 0 213 Connery 2002-08-17 965 495 216 MacBain 2011-09-10 392 0 216 MacBain 2009-08-30 388 0 225 Tompsen 2018-02-08 462 0 229 Willis 2019-07-18 643 179 229 Willis 2013-07-25 480 366 232 Cooper 2015-12-12 342 0 232 Cooper 2002-01-28 666 792 233 Williamson 2006-06-24 973 394 236 Smith 2012-11-04 561 0 238 Manning 2019-12-03 403 0 239 Booker 2014-05-20 828 476 240 Rowley 2016-11-17 416 0 241 Banks 2014-01-16 912 0 243 Ridgley 2016-10-13 22 0 250 Williamson 2019-02-03 335 1056 251 Barrister 2010-07-16 67 0 253 Tompsen 2016-12-15 448 0 253 Tompsen 2015-09-23 849 1003 256 Manning 2004-10-20 58 27 258 Presley 2019-02-25 52 0 260 Jones 2016-08-09 914 408 261 Maurer 2006-12-22 348 920 266 Goode 2007-04-27 74 0 266 Goode 2017-11-12 851 0 271 Tankersly 2003-06-13 381 781 273 Harper 2014-11-25 985 0 274 Booker 2018-04-14 783 404 275 Saylor 2007-12-19 634 0 281 Garth 2010-12-18 50 708 285 Peters 2019-11-04 428 0 285 Peters 2001-03-24 93 0 286 Cooper 2006-06-05 939 0 291 Dawson 2011-04-06 168 0 292 Peters 2011-04

max_ret_amnt 2006

- **How many customers have returns?**
    - There are 1024 customers that have returns.
- **What is the highest amount returned by a customer?**
    - The highest amount returned by a customer is 2006.
- **What is one unique thing about this query's results?**
    - One thing unique about this query result is that is shows the running total over time.

**Write a query to calculate the average bike price, the sum of all bikes sold, and the difference between the product price and sale price of every bike purchase. You can include any other columns you feel necessary.**

In [73]:
SELECT p.prod_id 
      ,p.prod_name 
      ,p.prod_price
      ,o.order_tot
      ,o.order_tot - p.prod_price AS sales_diff
      ,AVG(p.prod_price) OVER() as avg_bike_price
      ,SUM(o.order_tot) OVER(PARTITION BY p.prod_name) AS tot_bike_sales
FROM dsci_504.products AS p
INNER JOIN dsci_504.orders AS o
ON p.prod_id = o.prod_id;


(2599 row(s) affected)

Total execution time: 00:00:00.096

prod_id prod_name prod_price order_tot sales_diff avg_bike_price tot_bike_sales 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 4866.16 187.16 4587.7039746056175452 187747.24 49 AMD 4679 4880.2 201.2 4587.7039746056175452 187747.24 49 AMD 4679 4948.04 269.04 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 5006.53 327.53 4587.7039746056175452 187747.24 49 AMD 4679 4866.16 187.16 4587.7039746056175452 187747.24 49 AMD 4679 4941.02 262.02 4587.7039746056175452 187747.24 49 AMD 4679 5006.53 327.53 4587.7039746056175452 187747.24 49 AMD 4679 5018.23 339.23 4587.7039746056175452 187747.24 49 AMD 4679 5006.53 327.53 4587.7039746056175452 187747.24 49 AMD 4679 4983.14 304.14 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 5018.23 339.23 4587.7039746056175452 187747.24 49 AMD 4679 4905.93 226.93 4587.7039746056175452 187747.24 49 AMD 4679 5006.53 327.53 4587.7039746056175452 187747.24 49 AMD 4679 4936.35 257.35 4587.7039746056175452 187747.24 49 AMD 4679 4901.25 222.25 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 4887.22 208.22 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 4976.12 297.12 4587.7039746056175452 187747.24 49 AMD 4679 4948.04 269.04 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 5006.53 327.53 4587.7039746056175452 187747.24 49 AMD 4679 4901.25 222.25 4587.7039746056175452 187747.24 49 AMD 4679 4679 0 4587.7039746056175452 187747.24 49 AMD 4679 4983.14 304.14 4587.7039746056175452 187747.24 49 AMD 4679 4679 0 4587.7039746056175452 187747.24 49 AMD 4679 4901.25 222.25 4587.7039746056175452 187747.24 49 AMD 4679 5018.23 339.23 4587.7039746056175452 187747.24 49 AMD 4679 4983.14 304.14 4587.7039746056175452 187747.24 49 AMD 4679 4936.35 257.35 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 49 AMD 4679 4959.74 280.74 4587.7039746056175452 187747.24 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1937.88 87.88 4587.7039746056175452 81477.28 63 Air9 1850 1937.88 87.88 4587.7039746056175452 81477.28 63 Air9 1850 1951.75 101.75 4587.7039746056175452 81477.28 63 Air9 1850 1951.75 101.75 4587.7039746056175452 81477.28 63 Air9 1850 1984.13 134.13 4587.7039746056175452 81477.28 63 Air9 1850 1929.55 79.55 4587.7039746056175452 81477.28 63 Air9 1850 1850 0 4587.7039746056175452 81477.28 63 Air9 1850 1970.25 120.25 4587.7039746056175452 81477.28 63 Air9 1850 1924 74 4587.7039746056175452 81477.28 63 Air9 1850 1850 0 4587.7039746056175452 81477.28 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1967.48 117.48 4587.7039746056175452 81477.28 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1932.33 82.33 4587.7039746056175452 81477.28 63 Air9 1850 1970.25 120.25 4587.7039746056175452 81477.28 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1951.75 101.75 4587.7039746056175452 81477.28 63 Air9 1850 1979.5 129.5 4587.7039746056175452 81477.28 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1924 74 4587.7039746056175452 81477.28 63 Air9 1850 1977.19 127.19 4587.7039746056175452 81477.28 63 Air9 1850 1970.25 120.25 4587.7039746056175452 81477.28 63 Air9 1850 1924 74 4587.7039746056175452 81477.28 63 Air9 1850 1951.75 101.75 4587.7039746056175452 81477.28 63 Air9 1850 1961 111 4587.7039746056175452 81477.28 63 Air9 1850 1932.33 82.33 4587.7039746056175452 81477.28 63 Air9 1850 1850 0 4587.7039746056175452 81477.28 63 Air9 1850 1850 0 4587.7

- **What is one anomaly detected in the query results?**
    - There is only one order where there is a sales difference that is negative (-580.07). This may have occurred because of a sale, but it is interesting to see that it is the only order where the order total is significantly less than the product price.

In [77]:
/*Answering the second question (What was the total sales costs for Rocky Mountain bikes?): There aren't any Rocky Mountain Bikes, 
only components, so I have to approach the second question differently because the initial instructions lead to a different query. */

--Finding Rocky Mountain components
SELECT *
FROM dsci_504.components AS c 
WHERE c.comp_name LIKE '%Rocky Mountain%';

--Since it is a handlebar, we are going to search the builds to find which bikes contain the RM hbar (comp_id = 15).

SELECT *
FROM dsci_504.builds AS b
WHERE b.comp_hbar = '15';

--The build IDs that contain the RM hbar is 24 and 25.

SELECT o.*
      ,SUM(order_tot) OVER() AS tot_bike_sales
FROM dsci_504.orders AS o
WHERE o.build_id IN (24,25);




(1 row(s) affected)

(2 row(s) affected)

(75 row(s) affected)

Total execution time: 00:00:00.024

comp_id comp_name comp_cost comp_supplier comp_cat sup_id 15 Rocky Mountain AM 115 14 handlebar NULL

build_id comp_hbar comp_brake comp_wheels comp_fork comp_shock comp_shifter comp_cassette comp_derail comp_dropper comp_crank comp_chain_ring build_name 25 15 59 96 124 177 272 274 272 37 268 272 atomic 24 15 59 96 162 177 272 224 272 37 268 272 gnar buster

ord_id ord_date ord_tax_loc order_tot ord_ship_add ord_ship_date ord_track_num prod_id warehouse_id cus_id build_id tot_bike_sales 6921547 2012-02-26 23 5006.53 false 2012-03-06 PAC6665295-3 49 3 36 25 306729.24 5788991 2009-05-02 27 4880.2 false 2009-05-03 PAC7955013-2 49 1 110 25 306729.24 6920344 2017-01-19 9 3186.56 true 2017-01-25 PAC7907353-3 26 2 5 24 306729.24 5385043 2016-11-04 27 3195.75 false 2016-11-05 PAC9417722-0 26 3 19 24 306729.24 3513797 2001-04-16 16 4936.35 true 2001-04-22 PAC7002617-0 49 2 275 25 306729.24 4877732 2017-10-09 5 5018.23 true 2017-10-10 PAC7499398-1 49 3 334 25 306729.24 1378848 2011-01-30 13 4983.14 false 2011-02-09 PAC495926-0 49 3 345 25 306729.24 2419990 2004-08-19 21 3064 false 2004-08-21 PAC9463059-1 26 3 369 24 306729.24 3956821 2016-12-15 16 3232.52 false 2016-12-22 PAC606255-3 26 3 448 24 306729.24 3373348 2002-11-24 1 4679 false 2002-11-27 PAC314792-1 49 2 466 25 306729.24 6595312 2008-09-10 2 4866.16 false 2008-09-20 PAC4743331-0 49 1 491 25 306729.24 2411963 2016-07-06 21 3064 true 2016-07-11 PAC2945276-0 26 2 527 24 306729.24 1544577 2010-03-20 24 4959.74 false 2010-03-30 PAC8279433-3 49 3 618 25 306729.24 5117750 2000-09-17 19 3232.52 true 2000-09-24 PAC7970656-0 26 1 628 24 306729.24 5136786 2011-09-13 23 5006.53 false 2011-09-14 PAC3989529-0 49 3 676 25 306729.24 6485851 2016-09-25 20 4948.04 true 2016-10-02 PAC3415391-0 49 2 720 25 306729.24 6246934 2005-08-08 6 3258.56 false 2005-08-12 PAC6777599-3 26 3 850 24 306729.24 2525011 2000-03-07 8 4959.74 true 2000-03-11 PAC1089286-0 49 1 883 25 306729.24 6324619 2017-05-06 22 3064 true 2017-05-12 PAC5759034-0 26 2 905 24 306729.24 3611343 2005-05-11 12 5006.53 true 2005-05-11 PAC8724627-0 49 3 910 25 306729.24 2927778 2016-08-09 30 3186.56 false 2016-08-16 PAC6052237-2 26 3 914 24 306729.24 4834655 2017-07-26 26 4905.93 true 2017-08-01 PAC7319655-2 49 3 957 25 306729.24 7148611 2016-08-03 17 3274.65 false 2016-08-10 PAC632693-0 26 3 998 24 306729.24 3572324 2011-04-20 8 3247.84 false 2011-04-23 PAC2958313-0 26 2 1012 24 306729.24 4479302 2014-07-17 30 3186.56 false 2014-07-19 PAC8774258-1 26 2 1023 24 306729.24 3480213 2008-09-29 28 4983.14 false 2008-10-05 PAC1244694-1 49 1 1050 25 306729.24 1253281 2016-03-07 8 3247.84 true 2016-03-16 PAC4992841-3 26 1 1059 24 306729.24 6547105 2005-07-28 25 4959.74 true 2005-08-06 PAC5027734-3 49 1 1064 25 306729.24 5484128 2009-04-27 1 3064 true 2009-05-06 PAC1793135-3 26 3 1083 24 306729.24 5408819 2012-03-25 10 4866.16 false 2012-03-30 PAC441205-0 49 1 1097 25 306729.24 5833318 2017-06-13 17 3274.65 false 2017-06-17 PAC8972658-3 26 3 1100 24 306729.24 2528103 2018-02-08 8 4959.74 false 2018-02-12 PAC6271477-0 49 3 1121 25 306729.24 2992291 2001-12-06 14 4959.74 true 2001-12-08 PAC4592535-1 49 2 1130 25 306729.24 3271044 2017-11-04 29 3247.84 false 2017-11-05 PAC7560948-2 26 2 1133 24 306729.24 1265070 2009-01-04 1 4679 false 2009-01-05 PAC8940730-2 49 3 1178 25 306729.24 6507323 2000-06-22 17 3274.65 true 2000-07-02 PAC6613398-1 26 1 1183 24 306729.24 4371952 2012-10-25 12 3278.48 true 2012-10-30 PAC3504929-0 26 3 1198 24 306729.24 4675450 2011-06-22 29 4959.74 true 2011-07-02 PAC3432480-1 49 2 1234 25 306729.24 4556416 2012-06-14 17 3274.65 true 2012-06-23 PAC6892820-1 26 3 1373 24 306729.24 1778842 2011-08-18 21 3064 true 2011-08-22 PAC3745241-0 26 3 1401 24 306729.24 5732744 2001-04-26 24 4959.74 true 2001-04-27 PAC5156072-0 49 3 1449 25 306729.24 5867277 2010-08-02 6 3258.56 false 2010-08-07 PAC4430927-1 26 3 1473 24 306729.24 4191351 2019-04-19 29 4959.74 false 2019-04-24 PAC9568331-0 49 2 1477 25 306729.24 2399548 2010-10-15 3 4941.02 false 2010-10-18 PAC8204240-1 49 3 1500 25 306729.24 5807444 2001-03-12 23 5006.53 false 2001-03-20 PAC8406029-1 49 1 1505 25 306729.24 4138673 2004-05-31 18 4901.25 false 2004-06-06 PAC1828852-3 49 2 1573 25 306729.24 3944844 2007-04-02 15 4887.22 true 2007-04-12 PAC5960099-0 49 2 1603 25 306729.24 3623990 2017-10-08 20 4948.04 false 2017-10-13 PAC332018

- **What was the total sales costs for Rocky Mountain bikes?**
    - $306,729.24

**Part 6: Discuss your confidence in using advanced techniques such as window functions after completing this assignment. What advantages do they provide over aggregate functions?**

<span style="font-size: 16px;">I feel more confident about using windows functions after completing this assignment. I was a little nervous before working on this assingment because many people talk about how difficult windows functions can be, but they aren't too difficult. I actually prefer them over aggregate functions because of their advantages such as being able to retain the row information and being able to assign them row numbers, ranks, percentiles, etc.&nbsp; In additition to that, you can still use the aggregate functions (sum, count, min, max) as long as you use the OVER clause. Overall, windows functions are very convenient.</span>